cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -361,7 +395,34 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
         appMetadataStore.recordProgramDeprovisioning(programRunId, messageIdBytes);
         return Optional.of(provisioningService.deprovision(programRunId, datasetContext));
       case DEPROVISIONED:
-        appMetadataStore.recordProgramDeprovisioned(programRunId, endTs, messageIdBytes);
+        RunRecordMeta meta = appMetadataStore.recordProgramDeprovisioned(programRunId, endTs, messageIdBytes);
+        if (meta == null || meta.getCluster() == null || meta.getCluster().getNumNodes() == null) {
+          // this should not happen since when the cluster info should be there after deprovision
+          return Optional.empty();
+        }
+        Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(programRunId.getNamespaceId(),
+                                                                           meta.getSystemArgs());
+
+        if (!profile.isPresent()) {
+          // Each program run should be associated with a profile, so this should never happen
+          LOG.warn("Will not emit metrics for program {} without profile assigned to it, {}",
+                   programRunId, notification);
+          return Optional.empty();
+        }
+
+        ProfileId profileId = profile.get();
+        Integer node = meta.getCluster().getNumNodes();
+        // the node number will be null or 0 in local mode, set it to 1 so that we will just emit the difference
+        // between endts and starts
+        int nodeNum = node == null || node == 0 ? 1 : node;
```

Comment:
So from this comment https://github.com/caskdata/cdap/blob/develop/cdap-proto/src/main/java/co/cask/cdap/proto/ProgramRunCluster.java#L50, I thought this node number can still be null if it is on local cluster or it is not yet provisioned, when tested on sdk, I found actually it is 0, is there any situation where this will be null in some scenarios? If not I think I should update the comment

Reply:
that method can return null, but you check if it is null a few lines above, so when you get here it's not null

Reply:
oh right, I thought I added the null check when Intellij gave me some warning message, forgot to remove afterwards